Bluetooth: Host: Add support for Periodic Advertising with Responses (PAwR)

hermabe opened the pull request:
This PR adds host support for the Bluetooth 5.4 Core feature Periodic Advertising with Responses (PAwR). The API is a thin wrapper around the HCI commands and events used for the PAwR advertiser and scanner/sync.

Advertiser:
- Configure parameters
- Receive subevent data requests
- Set subevent data
- Receive response reports
   - Note: Reassembly of response reports are not implemented yet. I will do that in a followup PR.
  
Sync:
- Synchronize with a PAwR train using the same API as for Periodic Advertising Broadcast (Scan or PAST)
- Choose which subevents to synchronize to
- Receive advertising reports using the same API as for Periodic Advertising Broadcast
- Send responses

----
To test this you need a controller that supports PAwR advertiser and scanner/sync. The softdevice controller has support for this in the nrfconnect SDK. I have a work in progress sample here: https://github.com/hermabe/sdk-zephyr/commit/feb8f7b03b1fa7bf1a2532a5c55ef29c4d89506a that can be used to see usage of the API. I can upstream that sample if desired.

hermabe committed:
Samples: Bluetooth: Add samples for PAwR

Adds samples for both advertiser and sync.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Add support for PAwR Advertiser

Adds API for Periodic Advertising with Responses - Advertiser:
 - Configure parameters
 - Receive subevent data requests
 - Set subevent data
 - Receive response reports

The support is enabled by CONFIG_BT_PER_ADV_RSP, and requires
a controller that selects CONFIG_BT_CTLR_ADV_PERIODIC_RSP_SUPPORT.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe committed:
Bluetooth: Host: Add support for PAwR Sync

Adds API for Periodic Advertising with Responses - Scanner:
 - Synchronize to a PAwR train
 - Choose which subevents to synchronize to
 - Receive advertising reports from subevents
 - Send responses

The support is enabled by CONFIG_BT_PER_ADV_SYNC_RSP, and requires
a controller that selects CONFIG_BT_CTLR_SYNC_PERIODIC_RSP_SUPPORT.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

jori-nordic commented:
Could you add the sample to this PR as proposed?

hermabe commented:
> Could you add the sample to this PR as proposed?

Can do. It won't compile or run upstream though, as the zephyr controller doesn't support PAwR.

hermabe commented:
Added the sample. It can be tested with nrfconnect if it is cherry-picked to sdk-zephyr. Some configs also have to be set to enable the feature and set up the buffers.

alwa-nordic commented on include/zephyr/bluetooth/hci.h:
```diff
@@ -1488,6 +1499,56 @@ struct bt_hci_cp_le_ext_create_conn {
+struct bt_hci_cp_le_set_pawr_subevent_data {
+	uint8_t adv_handle;
+	uint8_t num_subevents;
+	struct bt_hci_cp_le_set_pawr_subevent_data_element subevents[0];
```
As I read it, the spec says the data is column-major. Pseudo code for illustration:
```suggestion
	uint8_t subevent[num_subevents];
	uint8_t response_slot_start[num_subevents];
	uint8_t response_slot_count[num_subevents];
	uint8_t subevent_data_length[num_subevents];
	uint8_t subevent_data[num_subevents];
```

hermabe replied:
The spec writes the layout a bit weirdly, but it is row-major. You can take a look at the extended create connection command with the initiating_phys for something similar.

alwa-nordic replied:
Discussed OOB. EBQ tests say it's in row-major. We'll assume that's correct.

alwa-nordic commented on subsys/bluetooth/host/adv.c:
```diff
@@ -1777,12 +1777,25 @@ static void adv_timeout(struct k_work *work)
 int bt_le_per_adv_set_param(struct bt_le_ext_adv *adv,
 			    const struct bt_le_per_adv_param *param)
 {
+#if defined(CONFIG_BT_PER_ADV_RSP)
+	struct bt_hci_cp_le_set_per_adv_param_v2 *cp;
+#else
 	struct bt_hci_cp_le_set_per_adv_param *cp;
+#endif /* CONFIG_BT_PER_ADV_RSP */
```
This must be an union. The check `BT_FEAT_LE_PAWR_ADVERTISER(bt_dev.le.features)` is runtime.

hermabe replied:
Added a comment.

hermabe commented on include/zephyr/bluetooth/bluetooth.h:
```diff
@@ -129,6 +162,32 @@ struct bt_le_ext_adv_cb {
+	 * @param adv     The advertising set object.
+	 * @param request Information about the upcoming subevents.
+	 */
+	void (*request)(struct bt_le_ext_adv *adv,
```
Will namespace them with `pawr`.

jori-nordic commented on subsys/bluetooth/controller/Kconfig:
```diff
@@ -607,6 +615,15 @@ config BT_CTLR_ADV_PERIODIC
+config BT_CTLR_ADV_PERIODIC_RSP
+	bool "LE Periodic Advertising with Responses in Advertising State"
+	depends on BT_BROADCASTER && BT_CTLR_ADV_PERIODIC_RSP_SUPPORT
+	select BT_CTLR_CHAN_SEL_2
```
missing `select EXPERIMENTAL`. same for SYNC

hermabe replied:
What would we be marking as experimental in that case? The zephyr controller does not implement PAwR and in sdk-nrf we are planning production support in the next release.

jori-nordic replied:
whoops I misread I though I was reading the host kconfig. It's all good then

hermabe commented on samples/bluetooth/periodic_adv_rsp/sample.yaml:
```diff
@@ -0,0 +1,9 @@
+tests:
+  sample.bluetooth.periodic_adv:
+    harness: bluetooth
+    platform_allow: qemu_cortex_m3 qemu_x86
```
I dont remember what we ended up with. If we want to move the sample to sdk-nrf or keep it here as host-only.

hermabe replied:
I will have to update the samples with correct main return as well.

jhedberg commented on samples/bluetooth/periodic_adv_rsp/src/main.c:
```diff
@@ -0,0 +1,388 @@
+{
+	printk("Connected (err 0x%02X)\n", err);
+
+	if (err) {
+		bt_conn_unref(conn);
```
This doesn't look correct. You don't own the `conn` reference in the connected callback, so you're not allowed to unref it either. If you want to keep a reference to it after the callback returns, then you need to take your own reference using `bt_conn_ref()`

alwa-nordic replied:
I agree. @hermabe The best approach is to tie the ownership of a reference to a variable. I just now made some wrapping functions around `bt_conn_ref`/`bt_conn_unref` that I think may be a safer interface.

```c
void conn_ref_store(struct bt_conn **conn_var, struct bt_conn *owned)
{
	void *conn_var_before = atomic_ptr_set((void **)conn_var, owned);
	__ASSERT_NO_MSG(conn_var_before == NULL);
}

struct bt_conn *__must_check conn_ref_move(struct bt_conn **conn_var)
{
	return atomic_ptr_clear((void **)conn_var);
}

void conn_ref_clone(struct bt_conn **conn_var, struct bt_conn *borrowed)
{
	struct bt_conn *owned = bt_conn_ref(borrowed);
	conn_ref_store(conn_var, owned);
}

void conn_ref_clear(struct bt_conn **conn_var)
{
	struct bt_conn *owned = conn_ref_move(conn_var);
	bt_conn_unref(owned);
}
```

Normal usage is trough `conn_ref_clone` and `conn_ref_clear`. `conn_ref_store` would only be used directly in special cases when you obtained an owned reference in a return value from somewhere, since you cannot take the address of a return value.

hermabe replied:
I have updated it to be more explicit. `bt_conn_le_create` gives a reference, which is unrefed in the failure cases. `conn` can never be `NULL` in the connected callback as all paths calling it dereference `conn`.

jhedberg commented on samples/bluetooth/periodic_adv_rsp/src/main.c:
```diff
@@ -0,0 +1,388 @@
+		return;
+	}
+
+	default_conn = conn;
```
Same here, i.e. this should read `default_conn = bt_conn_ref(conn);`

jhedberg replied:
Actually, how does this relate to setting of `default_conn` by `bt_conn_le_create()` on line 189? If the pointer is already set then, it seems you don't need to set it again in this function. That would however make calling `bt_conn_unref()` a valid thing in the error case, but in that case you should probably check that the pointer is non-NULL first (or is it even possible for it to be NULL?).

hermabe replied:
https://github.com/zephyrproject-rtos/zephyr/pull/56373#discussion_r1177556255

jhedberg commented on samples/bluetooth/periodic_adv_rsp/src/main.c:
```diff
@@ -0,0 +1,388 @@
+
+void disconnected(struct bt_conn *conn, uint8_t reason)
+{
+	if (conn != default_conn) {
```
Would it make sense to enforce `CONFIG_BT_MAX_CONN=1` for this sample, and then not have to do this extra checking?

jhedberg commented on samples/bluetooth/periodic_adv_rsp/src/main.c:
```diff
@@ -0,0 +1,388 @@
+
+void remote_info_available(struct bt_conn *conn, struct bt_conn_remote_info *remote_info)
+{
+	if (conn != default_conn) {
```
Same here.

hermabe committed:
Bluetooth: Host: Fix periodic sync reassembly

The reassembly buffer for periodic sync was not initialized
if the sync was established via PAST.

Move the initialization of the reassembly buffer to a common place.

Signed-off-by: Herman Berget <herman.berget@nordicsemi.no>

hermabe commented:
- Updated sample.yaml so that the samples are built with host-only in CI
- Set `CONFIG_BT_MAX_CONN=1`
- Removed some `bt_conn_ref` and `bt_conn_unref`
- Fixed an existing issue where reassembly of periodic advertising reports did not work when the sync was established via PAST.

hermabe commented:
Rebased on main and fixed yamllint.

jhedberg reviewed (changes requested):
One small issue, otherwise looks good to me

hermabe commented on samples/bluetooth/periodic_adv_rsp/src/main.c:
```diff
@@ -0,0 +1,381 @@
+	__ASSERT(conn == default_conn, "Unexpected connected callback");
+
+	if (err) {
+		bt_conn_unref(default_conn);
```
Updated.
